### PR TITLE
🚀 Add game settings

### PR DESCRIPTION
**Pull request by MaximeWrlz**
## Description

Add game settings (full working) in a dedicated modal)

## Checklist before requesting a review

-   [x] I have performed a self-review of my code
-   [x] I've added the corresponding label
-   [x] If my pull request implies some visual changes, I've linked screenshots in the description to show them
-   [x] If my pull request is linked to a task or a ticket, I've put the link in the description

--> [ClickUp](https://app.clickup.com/t/860rpd2ta)

## Screenshots

"Generals Parameters" tab:

Sans titre

"Weapons Selection" tab:

Sans titre 2png

### CODE REVIEW

**Comment by DrakenLine** [comment]
Task linked: [CU-860rpd2ta Games parameters](https://app.clickup.com/t/860rpd2ta)

File: lib/game-engine/OneVersusOne/game_page.dart [hunk 1]
```diff
@@ -101,12 +104,17 @@ class GamePage extends ConsumerWidget {
...
               ),
             const SizedBox(
               width: 30,
             ),
             if (gameState.gameSettings != null)
-              WeaponsList(weaponsList: gameState.gameSettings!.selectedWeapons)
+              WeaponsList(
+                  direction: Axis.vertical,
+                  isInGame: true,
+                  updateUI: () {},
```
Comment (DrakenLine) [inline lib/game-engine/OneVersusOne/game_page.dart:116]: maybe set `updateUI` as a non-required parameter ?

File: lib/game-engine/game.controller.dart [hunk 2]
```diff
@@ -212,6 +229,18 @@ class GameController extends StateNotifier<GameState> {
...
+  void updateSelectedWeapons(Weapon weapon) {
+    GameSettings gameSettings = state.gameSettings!;
+
+    if (gameSettings.selectedWeapons.contains(weapon)) {
+      gameSettings.selectedWeapons.remove(weapon);
+    } else {
+      gameSettings.selectedWeapons.add(weapon);
+    }
+
+    state = state..copyWith(updatedGameSettings: gameSettings);
```
Comment (DrakenLine) [inline lib/game-engine/game.controller.dart:241]:
```suggestion
    state = state.copyWith(updatedGameSettings: gameSettings);
```

File: lib/game-engine/game_engine.dart [hunk 3]
```diff
@@ -21,6 +21,23 @@ Map<String, dynamic> getGameJoiningSettings(String link, GamePlayer me) {
...
+    String instanceId, GameSettings gameSettings) {
+  List<String> weaponsNames = [];
+  for (var weapon in gameSettings.selectedWeapons) {
+    weaponsNames.add((weaponNameToString(weapon.name)));
+  }
+
+  return {
+    "data": {
+      "boardDimensions": gameSettings.boardDimensions,
+      "mode": 'one-versus-one',
```
Comment (DrakenLine) [inline lib/game-engine/game_engine.dart:34]:
```suggestion
      "mode": gameSettings.mode,
```

File: lib/pages/game/lobby_game/lobby_game.page.dart [hunk 4]
```diff
@@ -89,6 +90,41 @@ class LobbyGamePage extends ConsumerWidget {
                     mainAxisAlignment: MainAxisAlignment.center,
                     crossAxisAlignment: CrossAxisAlignment.center,
                     children: [
+                      Padding(
```
Comment (DrakenLine) [inline lib/pages/game/lobby_game/lobby_game.page.dart:93]: can you put it at the right of the start button ?

File: lib/services/socket.service.dart [hunk 5]
```diff
@@ -154,6 +155,13 @@ class WebSocketService {
           '[SOCKETS] ' 'Next turn : ${SocketsEventsListened.gameOver.value}',
         );
       });
+      socket.on(SocketsEventsListened.gameSettingsUpdated.value, (response) {
+        gameController.updateGameSettings(response['data']);
+        developer.log(
+          '[SOCKETS] '
+          'Game settings updated : ${SocketsEventsListened.gameSettingsUpdated.value}',
```
Comment (DrakenLine) [inline lib/services/socket.service.dart:162]:
```suggestion
          'Game settings updated',
```

File: lib/shared/Weapons/weapon_widget.dart [hunk 6]
```diff
@@ -66,12 +84,17 @@ class _WeaponWidgetState extends State<WeaponWidget> {
               children: [
                 Container(
                   foregroundDecoration: BoxDecoration(
-                    color: (widget.isLocked || !widget.isSelected)
-                        ? Colors.grey
-                        : Colors.transparent,
+                    color: ((widget.isInGame &&
```
Comment (DrakenLine) [inline lib/shared/Weapons/weapon_widget.dart:87]: I can barely read and understand it, can you make a function with distinct if ? 😅

File: lib/shared/gameConfig/game_config.dart [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
+        padding: const EdgeInsets.all(0.0),
+        child: Column(
+          children: <Widget>[
+            // Title + close button
+            Container(
+              width: double.infinity,
+              padding: EdgeInsets.only(
```
Comment (DrakenLine) [inline lib/shared/gameConfig/game_config.dart:75]:
```suggestion
              padding: EdgeInsets.symmetric(vertical : defaultPadding / 4, horizontal : defaultPadding),

```

File: lib/shared/gameConfig/game_config.dart [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+      GameController gameController = ref.read(gameProvider.notifier);
+      return Container(
+        decoration: BoxDecoration(
+          color: BSColors().purpleDarkBorder,
+        ),
+        child:
+            // Grid size
+            Padding(
+          padding: const EdgeInsets.all(20.0),
+          child: Column(
```
Comment (DrakenLine) [inline lib/shared/gameConfig/game_config.dart:176]: unnecessary Column

**Review by DrakenLine — CHANGES_REQUESTED** [review]

File: lib/game-engine/game.controller.dart (hunk 2 above)
Comment (MaximeWrlz) [reply lib/game-engine/game.controller.dart:241]: Weird, I wonder why it still worked

File: lib/shared/gameConfig/game_config.dart [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+              ),
+              decoration: BoxDecoration(
+                gradient: LinearGradient(
+                  colors: [BSColors().darkPurple, BSColors().purple],
+                  begin: Alignment.topCenter,
+                  end: Alignment.bottomCenter,
+                ),
+                borderRadius: const BorderRadius.only(
+                    topLeft: Radius.circular(6), topRight: Radius.circular(6)),
+              ),
```
Comment (AlexandreMichelix) [inline lib/shared/gameConfig/game_config.dart:89]: Apply the same decoration than the header widget :

```suggestion
              decoration: BoxDecoration(
                border: Border(
                    bottom:
                    BorderSide(color: BSColors().white.withOpacity(0.4), width: 2)),
                gradient: LinearGradient(
                  colors: [BSColors().darkPurple, BSColors().purple],
                  begin: Alignment.topCenter,
                  end: Alignment.bottomCenter,
                ),
              ),

```

File: lib/shared/gameConfig/game_config.dart [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+                    ),
+                  ),
+                  Consumer(builder: (context, ref, _) {
+                    GameController gameController = ref.read(gameProvider.notifier);
+                      return IconButton(
+                        padding: EdgeInsets.only(left: defaultPadding * 10),
+                        icon: Icon(Icons.close, color: BSColors().white),
+                        onPressed: () {
+                          resetGameSettings(gameController);
+                        },
```
Comment (AlexandreMichelix) [inline lib/shared/gameConfig/game_config.dart:111]: use ButtonWidget for the close button

**Review by AlexandreMichelix — APPROVED** [review]

**Pull request merged**